nexus-sdk/src/main/java/io/nexusrpc/handler/HandlerException.java

```diff
@@ -4,53 +4,124 @@
+     * behavior of the {@link HandlerException.ErrorType}.
+     */
+    NON_RETRYABLE
+  }
+
   private final ErrorType errorType;
+  private final RetryBehavior retryBehavior;
 
   public HandlerException(ErrorType errorType, String message) {
-    this(errorType, new RuntimeException(message));
+    this(errorType, new RuntimeException(message), RetryBehavior.UNSPECIFIED);
+  }
+
+  public HandlerException(ErrorType errorType, String message, RetryBehavior retryBehavior) {
+    this(errorType, new RuntimeException(message), retryBehavior);
   }
 
   public HandlerException(ErrorType errorType, @Nullable Throwable cause) {
+    this(errorType, cause, RetryBehavior.UNSPECIFIED);
+  }
+
+  public HandlerException(
+      ErrorType errorType, @Nullable Throwable cause, RetryBehavior retryBehavior) {
     super(cause == null ? "handler error" : "handler error: " + cause.getMessage(), cause);
     this.errorType = errorType;
+    this.retryBehavior = retryBehavior;
   }
 
   /** Error type for this exception. */
   public ErrorType getErrorType() {
     return errorType;
   }
 
+  /** Retry behavior for this exception. */
+  public RetryBehavior getRetryBehavior() {
+    return retryBehavior;
+  }
+
+  public boolean isRetryable() {
+    if (retryBehavior != RetryBehavior.UNSPECIFIED) {
+      return retryBehavior == RetryBehavior.RETRYABLE;
+    }
+    switch (errorType) {
+      case BAD_REQUEST:
+      case UNAUTHENTICATED:
+      case UNAUTHORIZED:
+      case NOT_FOUND:
+      case NOT_IMPLEMENTED:
+        return false;
+      case RESOURCE_EXHAUSTED:
+      case INTERNAL:
+      case UNAVAILABLE:
+      case UPSTREAM_TIMEOUT:
+        return true;
+      default:
+        throw new IllegalStateException("Unknown error type: " + errorType);
+    }
+  }
+
   /** Error type that can occur on a handler exception. */
   public enum ErrorType {
-    /** The server cannot or will not process the request due to an apparent client error. */
+    /**
+     * The server cannot or will not process the request due to an apparent client error. Clients
+     * should not retry this request unless advised otherwise.
+     */
     BAD_REQUEST,
-    /** The client did not supply valid authentication credentials for this request. */
+    /**
+     * The client did not supply valid authentication credentials for this request. Clients should
+     * not retry this request unless advised otherwise.
+     */
     UNAUTHENTICATED,
-    /** The caller does not have permission to execute the specified operation. */
+    /**
+     * The caller does not have permission to execute the specified operation. Clients should not
+     * retry this request unless advised otherwise.
+     */
     UNAUTHORIZED,
     /**
      * The requested resource could not be found but may be available in the future. Subsequent
-     * requests by the client are permissible.
+     * requests by the client are permissible but not advised.
```

Comment:
I'm changing this in go to
```
	// The requested resource could not be found but may be available in the future. Clients should not retry this
	// request unless advised otherwise.

```

Reply:
It's more consistent with the other non retryable errors.